### PR TITLE
Recursive and scoped RmDir.

### PR DESCRIPTION
**Pull request by mzhurovich**
@dkorolev PTAL!

### CODE REVIEW

**Comment by dkorolev** [comment]
LGTM!

Why not use an `enum class` instead of that `bool`, like we did for other parameters, with the default of "no"? Then, perhaps, the `*Recursive()` part of `RmDir*()` could use it too, and we'll have a backwards-compatible functionality extension that does not create any new names under `FileSystem::`?

Thanks,
Dima

**Comment by mzhurovich** [comment]
@dkorolev Like this?

File: Bricks/file/file.h [hunk 1]
```diff
@@ -252,11 +256,11 @@ struct FileSystem {
   }
 
   template <typename F>
-  static inline void ScanDir(const std::string& directory, F&& f) {
+  static inline void ScanDir(const std::string& directory, F&& f, bool include_dirs = false) {
```
Comment (dkorolev) [inline Bricks/file/file.h]: Here too, `s/bool/ScanRecursive::Yes/`. (I can't think of a better name now. :-( ideally, it's the same symbol that `RmDir` uses, with a more neutral name.)

File: Bricks/file/file.h [hunk 2]
```diff
@@ -284,21 +288,67 @@ struct FileSystem {
...
+          if (IsDir(full_name)) {
+            RmDir(full_name, parameters, RmDirRecursive::Yes);
+          } else {
+            RmFile(full_name, (parameters == RmDirParameters::ThrowExceptionOnError)
+                                  ? RmFileParameters::ThrowExceptionOnError
+                                  : RmFileParameters::Silent);
+          }
+        }, true);
+        RmDir(directory, parameters, RmDirRecursive::No);
+      } catch (...) {
```
Comment (dkorolev) [inline Bricks/file/file.h]: Sorry, but `catch (...)` is unacceptable in production code. I know right? ;-)

**Comment by dkorolev** [comment]
LGTM++

Please get rid of `catch(...)`, and sorry you had to suffer through the `clang-format` mess.

**Comment by mzhurovich** [comment]
@dkorolev PTAL once more, pls :)

File: Bricks/file/test.cc [hunk 3]
```diff
@@ -271,3 +273,58 @@ TEST(File, ScanDir) {
...
+  FileSystem::WriteStringToFile("1", f1.c_str());
+  FileSystem::MkDir(dir_y);
+  FileSystem::WriteStringToFile("2", f2.c_str());
+  FileSystem::MkDir(dir_z);
+  FileSystem::WriteStringToFile("3", f3.c_str());
+  FileSystem::WriteStringToFile("4", f4.c_str());
+
+  FileSystem::RmDir(dir_x, FileSystem::RmDirParameters::ThrowExceptionOnError, FileSystem::RmDirRecursive::Yes);
+
+  ASSERT_THROW(FileSystem::WriteStringToFile("test", f3.c_str()), FileException);
```
Comment (dkorolev) [inline Bricks/file/test.cc]: Any reason the order is `{ f3, f2, f1 }`? :-)

File: Bricks/file/test.cc [hunk 4]
```diff
@@ -271,3 +273,58 @@ TEST(File, ScanDir) {
...
+  {
+    const auto dir_remover = FileSystem::ScopedRmDir(dir_x);
+    FileSystem::MkDir(dir_x);
+    FileSystem::MkDir(dir_y);
+    FileSystem::MkDir(dir_z);
+    FileSystem::WriteStringToFile("PASSED", fn.c_str());
+    EXPECT_EQ("PASSED", FileSystem::ReadFileAsString(fn));
+  }
+  ASSERT_THROW(FileSystem::ReadFileAsString(fn), FileException);
+  ASSERT_THROW(FileSystem::RmDir(dir_z), DirDoesNotExistException);
```
Comment (dkorolev) [inline Bricks/file/test.cc]: Any reason the order is `{ dir_z, dir_y, dir_x }`? :-)

**Comment by dkorolev** [comment]
**LGTM++** : Final minor comments, feel free to merge this PR yourself after ignoring or taking care of them :-)

Thanks! :+1:

**Comment by mzhurovich** [comment]
@dkorolev Checks are in reverse order, in case something left inside the dirs during the process of recursive removal.

**Comment by dkorolev** [comment]
ACK!

**Pull request merged**